### PR TITLE
Allow user-defined filters

### PR DESCRIPTION
**Pull request by aleixalcacer**
Experimental work

### CODE REVIEW

File: blosc/blosc2.c [hunk 1]
```diff
@@ -886,6 +888,26 @@ uint8_t* pipeline_c(struct thread_context* thread_context, const int32_t bsize,
       case BLOSC_TRUNC_PREC:
         truncate_precision(filters_meta[i], typesize, bsize, _src, _dest);
         break;
+      case BLOSC_UDFILTER:
```
Comment (FrancescAlted) [inline blosc/blosc2.c:891]: Add a small comment like:

// Look for the user filter id in filters_meta and run it

File: blosc/blosc2.c [hunk 2]
```diff
@@ -1175,6 +1199,26 @@ int pipeline_d(blosc2_context* context, const int32_t bsize, uint8_t* dest,
       case BLOSC_TRUNC_PREC:
         // TRUNC_PREC filter does not need to be undone
         break;
+      case BLOSC_UDFILTER:
```
Comment (FrancescAlted) [inline blosc/blosc2.c:1202]: Ditto.

File: blosc/blosc2.h [hunk 3]
```diff
@@ -110,12 +110,14 @@ enum {
...
   BLOSC_TRUNC_PREC = 4,  //!< truncate precision filter
-  BLOSC_LAST_FILTER= 5,  //!< sentinel
+  BLOSC_UDFILTER = 5,  //!< user-defined filter
+  BLOSC_LAST_FILTER= 6,  //!< sentinel
 };
 
 enum {
   BLOSC2_MAX_FILTERS = 6,
   //!< Maximum number of filters in the filter pipeline
+  BLOSC2_MAX_UDFILTERS = 16,
```
Comment (FrancescAlted) [inline blosc/blosc2.h:120]: Add a comment here like:

// Maximum number of filters that a user can register.

File: blosc/blosc2.h [hunk 4]
```diff
@@ -775,7 +790,9 @@ typedef struct {
  */
 static const blosc2_cparams BLOSC2_CPARAMS_DEFAULTS = {
         BLOSC_BLOSCLZ, 5, 0, 8, 1, 0, NULL,
-        {0, 0, 0, 0, 0, BLOSC_SHUFFLE}, {0, 0, 0, 0, 0, 0},
+        {0, 0, 0, 0, 0, BLOSC_SHUFFLE},
+        {0, 0, 0, 0, 0, 0},
```
Comment (FrancescAlted) [inline blosc/blosc2.h:794]: The above line could be converted to `{0}` (for consistency with the line below).

File: blosc/blosc2.h [hunk 5]
```diff
@@ -764,6 +777,8 @@ typedef struct {
   //!< The (sequence of) filters.
   uint8_t filters_meta[BLOSC2_MAX_FILTERS];
   //!< The metadata for filters.
+  blosc2_udfilter udfilters[BLOSC2_MAX_UDFILTERS];
+    //!< The user-defined filters.
```
Comment (FrancescAlted) [inline blosc/blosc2.h:781]: Mind the indent space

File: blosc/blosc2.h [hunk 6]
```diff
@@ -789,12 +806,14 @@ typedef struct {
   //!< The number of threads to use internally (1).
   void* schunk;
   //!< The associated schunk, if any (NULL).
+  blosc2_udfilter udfilters[BLOSC2_MAX_UDFILTERS];
+    //!< The user-defined filters.
```
Comment (FrancescAlted) [inline blosc/blosc2.h:810]: Mind the indent space

File: examples/udfilters.c [hunk 7]
```diff
@@ -0,0 +1,171 @@
+/*
+  Copyright (C) 2015  Francesc Alted
```
Comment (FrancescAlted) [inline examples/udfilters.c:2]: Copyright should be:

Copyright (C) 2021  The Blosc Developers

File: examples/udfilters.c [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+  http://blosc.org
+  License: BSD 3-Clause (see LICENSE.txt)
+
+  Example program demonstrating use of the Blosc filter from C code.
```
Comment (FrancescAlted) [inline examples/udfilters.c:6]: The comments below does not belong here.  Better remove them.

**Review by FrancescAlted — COMMENTED** [review]
Looks very good.  Just did some small suggestions.

File: blosc/blosc2.c (hunk 1 above)
Comment (aalcacer) [reply blosc/blosc2.c:891]: Fixed

File: blosc/blosc2.c (hunk 2 above)
Comment (aalcacer) [reply blosc/blosc2.c:1203]: Fixed

File: blosc/blosc2.c (hunk 1 above)
Comment (aleixalcacer) [reply blosc/blosc2.c:891]: Fixed

File: blosc/blosc2.c (hunk 2 above)
Comment (aleixalcacer) [reply blosc/blosc2.c:1203]: Fixed

File: blosc/blosc2.h (hunk 3 above)
Comment (aleixalcacer) [reply blosc/blosc2.h:120]: Fixed

File: blosc/blosc2.h (hunk 6 above)
Comment (aleixalcacer) [reply blosc/blosc2.h]: Fixed

File: examples/udfilters.c (hunk 7 above)
Comment (aleixalcacer) [reply examples/udfilters.c]: Fixed

File: examples/udfilters.c (hunk 8 above)
Comment (aleixalcacer) [reply examples/udfilters.c]: Fixed

**Comment by FrancescAlted** [comment]
After some discussion, the current proposal is to drop the `BLOSC_UDFILTER` identifier and adopt the next convention:

1) Filter ID <= 32:  reserved for C-Blosc2
2) 32 < Filter ID <= 160: reserved for registered filters (TBD)
3) 160 < FilterID <= 255: reserved for user defined filters.

This convention will allow to use the corresponding byte in `filter_meta` for storing meta-information for *every* filter in C-Blosc2.

File: blosc/blosc2.h [hunk 9]
```diff
@@ -110,13 +125,10 @@ enum {
   BLOSC_BITSHUFFLE = 2,  //!< bit-wise shuffle
   BLOSC_DELTA = 3,       //!< delta filter
   BLOSC_TRUNC_PREC = 4,  //!< truncate precision filter
-  BLOSC_LAST_FILTER = 5,  //!< sentinel
+  BLOSC_LAST_FILTER = 6, //!< sentinel
+  BLOSC_LAST_REGISTERED_FILTER = 32,
```
Comment (FrancescAlted) [inline blosc/blosc2.h:129]: I suggest to mimic the same naming schema than for codecs (PR #261):

```
BLOSC_LAST_REGISTERED_FILTER = BLOSC2_BDEFINED_FILTERS + 0
```

File: blosc/blosc2.h [hunk 10]
```diff
@@ -110,13 +125,10 @@ enum {
   BLOSC_BITSHUFFLE = 2,  //!< bit-wise shuffle
   BLOSC_DELTA = 3,       //!< delta filter
   BLOSC_TRUNC_PREC = 4,  //!< truncate precision filter
-  BLOSC_LAST_FILTER = 5,  //!< sentinel
```
Comment (FrancescAlted) [inline blosc/blosc2.h:113]: Maybe you missed BLOSC_UDFILTER = 5 here?  (for consistency with #261)

File: blosc/blosc2.h [hunk 11]
```diff
@@ -1718,6 +1731,53 @@ BLOSC_EXPORT void blosc_set_blocksize(size_t blocksize);
...
+ * @brief The parameters for a user-defined filter.
+ */
+typedef struct {
+  uint8_t id;
+  //!< The filter identifier.
+  int (* forward)(const uint8_t *, uint8_t *, int32_t, blosc2_cparams *);
+  //!< The filter function that is used during compression.
+  int (* backward)(const uint8_t *, uint8_t *, int32_t, blosc2_dparams *);
+  //!< The filter function that is used during decompression.
+}blosc2_filter;
```
Comment (FrancescAlted) [inline blosc/blosc2.h:1769]: Add a blank space here.

File: tests/test_udfilters.c [hunk 12]
```diff
@@ -0,0 +1,245 @@
+/*
+  Copyright (C) 2020  The Blosc Developers
```
Comment (FrancescAlted) [inline tests/test_udfilters.c:2]: Minor: copyright is 2021

**Review by FrancescAlted — APPROVED** [review]
Looks good!  Just minor suggestions.

**Pull request merged**